src/main/java/org/commonjava/util/gateway/metrics/jfr/FlightRecorderFilter.java

```diff
@@ -0,0 +1,102 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.commonjava.util.gateway.metrics.jfr;
+
+import java.io.IOException;
+
+import javax.ws.rs.container.ContainerRequestContext;
+import javax.ws.rs.container.ContainerRequestFilter;
+import javax.ws.rs.container.ContainerResponseContext;
+import javax.ws.rs.container.ContainerResponseFilter;
+import javax.ws.rs.ext.Provider;
+
+import org.commonjava.util.gateway.metrics.jfr.events.JaxRSEvent;
+import org.jboss.resteasy.core.ResourceMethodInvoker;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Purpose of this filter is to generate events for JaxRS calls when profiling is enabled.
+ *
+ */
+@Provider
+public class FlightRecorderFilter implements ContainerRequestFilter, ContainerResponseFilter
+{
+
+    @Override
+    public void filter( ContainerRequestContext requestContext ) throws IOException
+    {
+        logger.trace("processing request context");
+        final JaxRSEvent event = new JaxRSEvent();
+        final boolean isEnabled = event.isEnabled();
+        if ( !isEnabled )
+        {
+            return;
+        }
+        event.begin();
+        requestContext.setProperty( JaxRSEvent.NAME, event );
+    }
+
+    @Override
+    public void filter( ContainerRequestContext requestContext, ContainerResponseContext responseContext )
+            throws IOException
+    {
+        logger.trace("processing response context");
+        if ( requestContext == null )
+        {
+            logger.error( "request context is null" );
+            return;
+        }
+        Object prop = requestContext.getProperty( JaxRSEvent.NAME );
+        if ( prop == null )
+        {
+            return;
+        }
+        JaxRSEvent event = ( JaxRSEvent ) prop;
+        if ( !event.isEnabled() )
+        {
+            return;
+        }
+
+        event.end();
+
+        if ( event.shouldCommit() ) {
+            event.method = requestContext.getMethod();
+            event.mediaType = String.valueOf ( requestContext.getMediaType() ) ;
+            event.length = requestContext.getLength();
+            event.methodFrameName = getMethodName( requestContext );
+            event.path = requestContext.getUriInfo().getPath();
+            event.responseLength = responseContext.getLength();
+            event.status = responseContext.getStatus();
+
+            event.commit();
+        }
+    }
+
+    private String getMethodName( ContainerRequestContext context )
+    {
+        Object p = context.getProperty( METHOD_NAME );
+        if ( p == null )
+        {
+            return "";
+        }
+        ResourceMethodInvoker invoker = (ResourceMethodInvoker) p;
```

Comment:
I guess this is how containerreqcontext works. do we add a class cast check here?

Reply:
Yes good idea.